### PR TITLE
Refactor transfer funds loading 

### LINKED ISSUE #3969: TransferFunds loads forever if no safe selected
## Steps to reproduce

Go to safe control
Don't select a safe
Select transfer funds

## Expected behaviour

You get an error message telling you to select a safe like in https://github.com/JoinColony/colonyDapp/pull/3926

## Actual behaviour

Loads forever

loadforever

### PR DESCRIPTION
**Pull request by willm30**
**Note, only the first commit belongs uniquely to this PR. The others are from a merge of another PR**

## Description

This PR refactors the loading pattern in the transfer funds section. 

First, it ensures there's an appropriate error message if the safe isn't selected. 
Second, it implements a cache, so that we only fetch for token data once per safe.  
Third, it abstracts the `loading` and `error` message elements so that they can be used in other sections ([#3998](https://github.com/JoinColony/colonyDapp/pull/3993/commits/950bf41b69bce90802f3502f23bcd4c6f0c37423)).

**Changes** 🏗

* Safe not selected message
* Remember previously fetched tokens
* Abstracted loading and error messages

Resolves #3969

### CODE REVIEW

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/validation.ts [hunk 1]
```diff
@@ -63,8 +63,7 @@ const MSG = defineMessages({
   },
   balanceError: {
     id: `dashboard.ControlSafeDialog.validation.balanceError`,
-    defaultMessage:
-      'Could not retreive balance information. Please try again later',
+    defaultMessage: 'Could not retreive balance information',
```
Comment (jakubcolony) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/validation.ts:66]: Little typo here

**Review by jakubcolony — APPROVED** [review]
Nice work @willm30!

image

image


image

I noticed this weird behaviour of "Max" button which I don't think was caused by this PR but might still be worth looking into:

https://user-images.githubusercontent.com/112586815/196540137-672249a9-fcd9-4871-ba81-b132b3fc8263.mov

**Comment by jakubcolony** [comment]
>I noticed this weird behaviour of "Max" button which I don't think was caused by this PR but might still be worth looking into:

In fact, I just encountered this again while reviewing #3977. I created an issue for it - #4011.

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/validation.ts (hunk 1 above)
Comment (willm30) [reply src/modules/dashboard/components/Dialogs/ControlSafeDialog/validation.ts:67]: Thanks, will update before merging

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/TransactionTypesSection/TransferFundsSection.tsx [hunk 2]
```diff
@@ -41,67 +42,107 @@ const MSG = defineMessages({
...
+  <DialogSection>
+    <div className={styles.error}>
+      {typeof error === 'string' ? (
+        <span>{error}</span>
+      ) : (
+        <FormattedMessage {...error} />
+      )}
+    </div>
+  </DialogSection>
+);
```
Comment (ArmandoGraterol) [inline src/modules/dashboard/components/Dialogs/ControlSafeDialog/TransactionTypesSection/TransferFundsSection.tsx:88]: Are these leftovers from a rebase? I remember that they were supposed to be extracted for all sections to use, right?

**Review by ArmandoGraterol — APPROVED** [review]
LGTM :ok_hand:

File: src/modules/dashboard/components/Dialogs/ControlSafeDialog/TransactionTypesSection/TransferFundsSection.tsx (hunk 2 above)
Comment (willm30) [reply src/modules/dashboard/components/Dialogs/ControlSafeDialog/TransactionTypesSection/TransferFundsSection.tsx:88]: I do it in this one https://github.com/JoinColony/colonyDapp/pull/3998/commits/dd38f34fdace72c4cbb6397b1555cb70220aebef#diff-d9b918ffd427089e5685173ef3ff58bfd6e8016420e3fa43120248e5fe19b88c

**Pull request merged**